src/SLCore/Core/ISLCoreListenerSetUp.cs

```diff
@@ -0,0 +1,53 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+
+namespace SonarLint.VisualStudio.SLCore.Core
+{
+    public interface ISLCoreListenerSetUp
+    {
+        void Setup();
+    }
+
+    [Export(typeof(ISLCoreListenerSetUp))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class SLCoreListenerSetUp : ISLCoreListenerSetUp
+    {
+        private readonly ISLCoreJsonRpc wrapper;
+        private readonly IEnumerable<ISLCoreListener> listeners;
+
+        [ImportingConstructor]
+        public SLCoreListenerSetUp(ISLCoreJsonRpc wrapper, [ImportMany] IEnumerable<ISLCoreListener> listeners)
```

Comment:
What happens when we need to update the wrapper (when core restarts)? We need to re-register all listeners with this wrapper. It needs to be moved to arguments of Setup method

Reply:
If you run the setup again it will just attach them to the wrapper again why do we need to have them as a parameter of the method?

Reply:
The wrapper is not going to be mef-importable